### ISSUE TITLE
Improving Release process

### ISSUE THREAD

**Issue by AndreyNikiforov** [icloud-photos-downloader/icloud_photos_downloader#181]
Current process is a set of manual steps -- automating these will reduce chances of errors and reduce inconsistencies: today we have 1.6.0 release version in codebase, 1.6.1 in PyPy, 1.3.4 in GitHub, `latest` in dockerhub with 1.6.0

Current release process is performed by @ndbroadbent (Owner; the only one with access to pypy and dockerhub), creating potential delays for releases. Distributing this role among Maintainers will reduce dependencies.

I am proposing to use github actions for creating .whl assets, publishing PyPy packages, and pushing Docker images to dockerhub on each gitHub Release publishing

Benefits:
- All credentials to PyPy and Dockerhub remain in the hands of Owner (but have to be added to the repo secrets)
- Github Releases can be created by Maintainers, producing releases at any schedule
- Process is automated, consistent, and repeatable

Additionally, we can use github action for testing pull requests/commits instead of travis. Travis account will no longer be needed

Not changing:
- coverall

Challenges:
- implementing/testing gitlab actions may be challenging without publishing new versions

**Comment by menkej** [comment]
I'm really interested in this. I'm having similar issues at my "day job", where we're just starting to get near to what you describe @AndreyNikiforov. It makes totally sense for me. I have no glue of the publishing processes of these platforms but I'd be interested to dive into it.

**Comment by menkej** [comment]
> we can use github action for testing pull requests/commits

I created a GitHub action for running the tests. Seems to work. Feel free to mess around with it. ;-)

**Comment by AndreyNikiforov** [comment]
> > we can use github action for testing pull requests/commits
> 
> I created a GitHub action for running the tests. Seems to work. Feel free to mess around with it. ;-)

@menkej That is very nice work! Follow up questions and suggestions:
- Actions seems to be failing for `master` now. Can you see if they can be fixed?
- It may be easier to iterate on your own fork and then submit PR
- Do you think we can use testing on `push`es to every branch, not only `master`? That way workflow will work in forks and help maintaining quality before submitting PRs.

**Comment by AndreyNikiforov** [comment]
@menkej another ideas provoked by your testing action: use matrix strategy to test on different python versions (https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions/managing-complex-workflows#using-a-build-matrix)

**Comment by menkej** [comment]
> > > we can use github action for testing pull requests/commits
> > 
> > 
> > I created a GitHub action for running the tests. Seems to work. Feel free to mess around with it. ;-)
> 
> @menkej That is very nice work! Follow up questions and suggestions:
> 
> * Actions seems to be failing for `master` now. Can you see if they can be fixed?

I did not dive into it yet. @ndbroadbent committed two things and now one test case is failing:
`FAILED tests/test_download_photos.py::DownloadPhotoTestCase::test_until_found
252
`
> * It may be easier to iterate on your own fork and then submit PR

Sure, this is the way I work normally. I did not know I could do this with actions, too... I just noticed that these are forked as well. Nice.

The action can be found here: `.github/workflows/quality-checks.yml`

> * Do you think we can use testing on `push`es to every branch, not only `master`? That way workflow will work in forks and help maintaining quality before submitting PRs.

That's exactly what I intended. Maybe the trigger event needs some tweaking. Just started diving into Actions (and GitHub in general). Used ClearCase and CVS/PVCS for ages - still a git greenhorn... ;-)

**Comment by menkej** [comment]
> use matrix strategy

Already on my wishlist. ;-) Currently the tests run under Linux, python 3.8. Mac/Win and some python (eg 3.6, as this ist minimal requirement AFAIR) variance might make sense.

**Comment by AndreyNikiforov** [comment]
> I did not dive into it yet. @ndbroadbent committed two things and now one test case is failing:
> `FAILED tests/test_download_photos.py::DownloadPhotoTestCase::test_until_found 252 `

I noticed that `test_until_found` is temperamental in general. Had hard time making it working on Windows without success.

**Comment by AndreyNikiforov** [comment]
> > use matrix strategy
> 
> Already on my wishlist. ;-) Currently the tests run under Linux, python 3.8. Mac/Win and some python (eg 3.6, as this ist minimal requirement AFAIR) variance might make sense.

Awesome. Let's start with different python version on Linux, then tackle on releases, and only then add more platforms to the mix. Rationale is that platform beyond Linux are usually second class citizens in CI world and we may have challenges that we are not ready for yet. One of them is that tests may not work on all platforms -- I know Windows tests are limited now.

**Comment by menkej** [comment]
I have absolutely no idea of the releasing process. It is rather easy to set this up if you know what has to be done. I'm still learning python and all this stuff, so I'd need some assistance here. @ndbroadbent  maybe we can talk about what you're usually doing for a release and I'll take care of the "auto-magic" then?

**Comment by AndreyNikiforov** [comment]
@menkej here is rough plan that I have:
1. use test personal project for experimentation
1. add `release` action (to test project)
    1. look at `scripts` folder in this project for steps to replicate (test, publish to coverall, build whl, push to PyPi, build docker, push dockerhub)
    1. add keys for Pypi, dockerhub, coverall to GitHub secrets
1. once the process learnt and tested
    1. ask for @ndbroadbent to add necessary keys to this repo
    1. `port` release work from test repo to this repo
    1. clean test project (remove dockerhub, Pypi, coverall)
1. test release process in this repo using `rc` or `alpha` release

**Comment by menkej** [comment]
I have upload to pypi working - triggered by the creation of a release.

I was playing around with an action that creates the release automatically - including release notes from git commits - when a new v123.45.6 tag is created. I have some issues with the trigger event, but the content should be working.

Overall: initial creation of Actions really sucks, as I could not really find a way to create an Action on a branch (for clean pull request) from the web interface. Maybe I try it completely "by hand", at least for the first step. But not today... ;-)

**Comment by menkej** [comment]
As I did not get a reaction to this post in #198 (and this is closed now) I copied it here... So @ndbroadbent @AndreyNikiforov should we this give a try? But then we need the secrets added into this projects configuration....
---
Ok, finally... Actions can be quite a beast... I have created an action that works as follows:

- you edit CHANGELOG  and add a new RELEASE. (lets say 1.7.0 for now --> following SemVer of course.) The structure stays as
```
## Unreleased

## 1.7.0 (2020-11-1)

- fix: --log-leve...
```
- you edit setup.py to reflect the new RELEASE version 1.7.0
- you commit
- you add a tag "v1.7.0" - question: is there a possible security flaw maybe? Or is creation of tags just possible for those with write access to origin?
- you push --tag
- This triggers the action(s). Quality checks are run and the new action "Create Release". This covers steps 2,3,4 mentioned above.:
```
2. build .whl (name with version number)
3. publish .whl to PyPI
4. create github release with notes from Changelog and .whl as asset; name of the release is version number
```

- publishing is currently to testpypi-repository, as we need the correct secrets in GitHub-Secrets

**For review:**
1. You can check the action here: https://github.com/menkej/icloud_photos_downloader/blob/master/.github/workflows/createrelease.yml
1. You can find a run of that action here: https://github.com/menkej/icloud_photos_downloader/actions
1. You can find the generated release here: https://github.com/menkej/icloud_photos_downloader/releases/tag/v1.7.19 (Version numbers not matching as I needed several runs to make it work...). If the credentials for docker are added to The GitHub-Secrets as well it will be quite easy to add the docker steps, too.

Maybe it could be possible to automate setup.py-change

**Learnings and questions:**
- working with Actions requires working on `master`. Adding an action via a branch/pull request did not work for me. Is this possible? Modifying should work, though, but I'm not sure about testing then....
- debugging Actions is "different" and works best via pushing/testing online. Setting `ACTIONS_RUNNER_DEBUG`and `ACTIONS_SETP_DEBUG` helps to get additional info (online).
- Local testing of actions, p.e. with `act` does not cover all the necessary things - at least I did not get this Workflow to run
- By working on `master` I messed up with my fork. What is the best practice to clean that up and get back in sync with upstream? The easiest would probably be just to delete the action, but that is probably not the "right" way someone would do that in git.

What do you think, guys?

**Comment by ndbroadbent** [comment]
Sure this all sounds great! I can definitely set up a new user for PyPi and Docker Hub and share the password in GitHub secrets. Please let me know how I can do that, and I can get that set up

**Comment by menkej** [comment]
> I can definitely set up a new user for PyPi and Docker Hub and share the password in GitHub secrets. Please let me know how I can do that, and I can get that set up. My current script uses `PYPI_USERNAME` and `PYPI_PASSWORD` secrets. Secrets are set up in `Settings/Secrets`. Thats it. Then I can create the action on master use it. I have not implemented docker hub yet but that is more or less copy&paste from the existing script. I suggest to use `DOCKERHUB_USERNAME` and `DOCKERHUB_PASSWORD` (very creative, I know... ;-) ) here...

Btw: does this project have an email-address?

**Comment by ndbroadbent** [comment]
Oh great! Sure I will create the new users and give them the permissions, then set up these variables on the secrets page. I'll let you know when that's done! (Probably tomorrow morning since I'm about to head offline now)

**Comment by ndbroadbent** [comment]
Oh also I haven't registered any domains or email addresses for the project

**Comment by ndbroadbent** [comment]
Hey @menkej, I've created some new users, given them the required permissions, and set up the GitHub secrets for `PYPI_USERNAME`, `PYPI_PASSWORD`, `DOCKERHUB_USERNAME`, `DOCKERHUB_PASSWORD`. So you should be able to use these usernames and passwords to sign in and push new PyPi packages and Docker images.

<img width="930" alt="Screen Shot 2020-11-04 at 10 05 37 PM" src="https://user-images.githubusercontent.com/139536/98121153-e0a42880-1ee9-11eb-9557-747f9cd1aa5f.png">

Please let me know if you run into any issues with the permissions!

**Comment by ndbroadbent** [comment]
Oh also I did need to provide an email address for both of these new accounts, so I've used: `icloudpd <at> ndbroadbent.com`

**Comment by AndreyNikiforov** [comment]
> Oh also I did need to provide an email address for both of these new accounts, so I've used: `icloudpd <at> ndbroadbent.com`

What is the common practice for email for open source projects? I could think of distribution list with management access for a number of owners. May be GitHub has something already?

**Comment by menkej** [comment]
> Hey @menkej, I've created some new users, given them the required permissions, and set up the GitHub secrets for `PYPI_USERNAME`, `PYPI_PASSWORD`, `DOCKERHUB_USERNAME`, `DOCKERHUB_PASSWORD`. So you should be able to use these usernames and passwords to sign in and push new PyPi packages and Docker images.

Excelllent! I hope I'll find some time tomorrow or Saturday morning to update the Action so that - maybe - we could try the new process on this weeks release already - starting with an rc-release, of course!

**Comment by menkej** [comment]
Hey guys,
my wife had other plans, so I spend some hours fighting with my "create release" action. This is a love-hate relationship already... I have an issue with the docker build that I have not found yet, but for today I'm "done". I ended up with way to many tabs with GitHub-logo favicon... ;-)

Maybe one of you two could have a look at the log and has an idea...

- Log: https://github.com/menkej/icloud_photos_downloader/runs/1365714804?check_suite_focus=true
- I took inspiration from here: https://github.com/marketplace/actions/build-and-push-docker-images
- to add to my action: https://github.com/menkej/icloud_photos_downloader/blob/master/.github/workflows/createrelease.yml

Another "issue" I had: I wanted to update setup.py/CHANGELOG and tag addition in one push. Is it possible that `git push --tags` pushes *only* tags and not the other commit? At least the action is not triggered if I do it at once. So currently two step: change the files. commit. push. Add tag. push --tags. -> The magic begins. Not sure if I'm really happy with this.

A new thing I added on the way: check if the tag name matches the version in setup.py and if not "do nothing". todo: this should lead to an error.

I need a drink and a Netflix-reset... ;-)

**Comment by menkej** [comment]
Ok, I'm one step further. The Dockerfile.release depends on a dist directory which seems not to be present - it should be there after a normal build with `scripts/build` but maybe I'm just lost in virtualisation layers... Hints for a Docker noob?

Possible solution: push to pypi first and install icloudpd via pip then... As I'm currently pushing to test.pypi I need to adapt...

**Comment by boredazfcuk** [comment]
> Ok, I'm one step further. The Dockerfile.release depends on a dist directory which seems not to be present - it should be there after a normal build with `scripts/build` but maybe I'm just lost in virtualisation layers... Hints for a Docker noob?
> 
> Possible solution: push to pypi first and install icloudpd via pip then... As I'm currently pushing to test.pypi I need to adapt...

AFAIK it's best practice to put everything in a single layer. Personally I've found that something performed in one layer doesn't apply to the next.

You probably want your Dockerfile.release to look something like this:
```
FROM alpine:latest
ARG ICLOUDPD_VERSION
COPY dist/* /tmp
RUN set -xe && \
   adduser -D -h /home/user -u 1000 user && \
   apk add --no-cache python3 && \
   python3 -m ensurepip && \
   pip3 install --upgrade pip setuptools && \
   pip install /tmp/icloudpd-${ICLOUDPD_VERSION}-py2.py3-none-any.whl && \
   icloudpd --version && \
   icloud -h | head -n1 && \
   if [ ! -e /usr/bin/pip ]; then ln -s pip3 /usr/bin/pip ; fi && \
   if [[ ! -e /usr/bin/python ]]; then ln -sf /usr/bin/python3 /usr/bin/python; fi && \
   rm -r /usr/lib/python*/ensurepip && \
   rm -r /root/.cache
```

**Comment by AndreyNikiforov** [comment]
just ran `scripts/build` on mac and dist folder with whl files was created there. `scripts/build_docker` did not work on mac as `realpath` was not found, but that is separate issue.

**Comment by AndreyNikiforov** [comment]
> You probably want your Dockerfile.release to look something like this:

Good idea @boredazfcuk ! Let's put this optimization as its own PR (separate from release process improvements)!

**Comment by AndreyNikiforov** [comment]
> Another "issue" I had: I wanted to update setup.py/CHANGELOG and tag addition in one push. Is it possible that `git push --tags` pushes _only_ tags and not the other commit? At least the action is not triggered if I do it at once. So currently two step: change the files. commit. push. Add tag. push --tags. -> The magic begins. Not sure if I'm really happy with this.

@menkej do you think we can automate what we have first and then optimize and enhance the process?

**Comment by menkej** [comment]
> just ran `scripts/build` on mac and dist folder with whl files was created there. `scripts/build_docker` did not work on mac as `realpath` was not found, but that is separate issue.

I had faked realpath on Mac for the scripts I'm using by putting this above in the scripts...

```
# realpath not availabe on Mac by default...
realpath() {
    [[ $1 = /* ]] && echo "$1" || echo "$PWD/${1#./}"
}
```

@ndbroadbent did introduce that just recently (commit 39c0648). Dunno why.

**Comment by menkej** [comment]
@AndreyNikiforov / @boredazfcuk you're obviously more into Docker than me (which is not hard). I spent a very long time this weekend trying to get the release fully automated but fail on docker using the recommended way (https://github.com/marketplace/actions/build-and-push-docker-images#git-context).

I also tried using @boredazfcuk version of `Dockerfile.release` and separated the RUNs. It seems like buildx runs several builds on different platforms in parallel and I don't really get what is happening here. Locally (Mac, Docker CE) I can build the Docker image using `scripts/build && scripts/build_docker`. My action basically does the same but the Docker part fails.

Maybe you can give me a push if you look at https://github.com/menkej/icloud_photos_downloader/runs/1371090173?check_suite_focus=true ?

**Comment by boredazfcuk** [comment]
The `buildx` Docker plugin is for building multi-architecture builds. You can specify which architectures you want by using the --platform option, so something like:
```
docker buildx build . --platform "linux/i386,linux/amd64,linux/arm64,linux/arm/v7,linux/arm/v6" --tag icloudpd/icloudpd
```
adding ` --push` will push the image up to Dockerhub.

The native `docker build`  command (invoked in Dockerfile.release) will only build the image for the host machine's current architecture, presumably amd64. If you're pushing to Dockerhub, you're going to want to focus on getting the buildx version working.

Looking at the error, I see `exec user process caused: exec format error`

Nailed on, this is going to be an end-of-line character encoding issue. Windows uses <Carriage Return> + <Line Feed> to signify the end of a line. Linux only uses <Line Feed>. I believe Mac uses <Carriage Return> only, but I don't use Mac so can't say for sure. When a Windows text file is copied to a Linux platform, you'll see a  `^M` character on the end of each line which is the representation of the <Carriage Return> character. Pretty sure this is why it's complaining about a format error.

I have a https://github.com/boredazfcuk/docker-icloudpd/blob/master/.gitattributes file which tells git that any .sh files that it pulls down should change the end-of-line character to <Line Feed>:
```
*.sh text eol=lf
```

I'm not 100% sure which file is going to be causing it that results in the error, but it's going to be whichever one has the `ls /tmp` command in it. Try adding that file into a `.gitattributes` file, pulling the repo and running the build again.

**Comment by menkej** [comment]
> Looking at the error, I see `exec user process caused: exec format error`
> 
> Nailed on, this is going to be an end-of-line character encoding issue.

Thanks for the guess. I didn't really expect it to change, but I tried adding `.gitattributes`. As you can see in the current run (https://github.com/menkej/icloud_photos_downloader/runs/1371289765?check_suite_focus=true) the error appeared again (lines 177ff):

```
#9 [linux/arm64  3/14] RUN ls /tmp
#9 0.120 standard_init_linux.go:219: exec user process caused: exec format error
#9 ERROR: executor failed running [/bin/sh -c ls /tmp]: exit code: 1
#37 [linux/386  3/14] RUN ls /tmp
#37 0.144 icloudpd-1.7.17rc4-py2-none-any.whl
#37 DONE 0.2s
```

What makes it more strange is that the "ls" works in linux/386 but fails on linux/arm64? I have absolutely no idea.

The command comes from my patched `Dockerfile.release`. Initially I had issues that the whl-file was not found as the version was not populated correctly. Anyways: the action is still failing, and I have no Idea why. This is why I ended up using a different `Dockerfile.release` (inspired by your suggestion) and separated the RUNs.

Current `Dockerfile.release`:

```
FROM alpine:latest
ARG ICLOUDPD_VERSION
COPY dist/* /tmp
RUN ls /tmp
RUN adduser -D -h /home/user -u 1000 user
RUN apk add --no-cache python3
RUN python3 -m ensurepip
RUN pip3 install --upgrade pip setuptools
RUN pip install /tmp/icloudpd-${ICLOUDPD_VERSION}-py2.py3-none-any.whl
RUN icloudpd --version
RUN icloud -h | head -n1
RUN if [ ! -e /usr/bin/pip ]; then ln -s pip3 /usr/bin/pip ; fi
RUN if [[ ! -e /usr/bin/python ]]; then ln -sf /usr/bin/python3 /usr/bin/python; fi
RUN rm -r /usr/lib/python*/ensurepip
RUN rm -r /root/.cache
```

I can build that locally on Mac without issues with `docker buildx build --build-arg ICLOUDPD_VERSION=1.7.17rc4 --tag icloudpd/icloudpd:latest --tag icloudpd/icloudpd:1.7.17rc4 --file ./Dockerfile.release .`. The only difference from the run via GitHub is that I omitted `--platform linux/amd64,linux/arm64,linux/386 --iidfile /tmp/docker-build-push-attQNR/iidfile` as multiple platforms seem not to be supported locally.

**Comment by AndreyNikiforov** [comment]
Some notes on `buildx`:
- it is an experimental technology
- it comes with Docker Desktop but not with docker ce images used in CI/CD
- building multiarch images requires arm support, e.g. with qemu (true for regular builds for arm too)
- pushing multiarch images with buildx to some repositories is problematic, but I do not know about Dockerhub
- builds are faster and from my experience are safe replacement for local builds

With that said, let's focus on making regular `docker build` working in automatic way in GitHub and then work on making multiarch images (and buildx) in GitHub actions.

**Comment by AndreyNikiforov** [comment]
> Looking at the error, I see `exec user process caused: exec format error`

That is most likely error due to binary image format problems -- arm on amd platform without emulator.

**Comment by menkej** [comment]
> > Looking at the error, I see `exec user process caused: exec format error`
> 
> That is most likely error due to binary image format problems -- arm on amd platform without emulator.

Thanks Andrey! Sometimes... I had commented out Qemu for testing in the action. Reenabled and got further. Now I ran into the next issue: Script expects a different filename than the build provides.... I'll dig into that tomorrow.

see the run here: https://github.com/menkej/icloud_photos_downloader/runs/1371448214?check_suite_focus=true

**Comment by menkej** [comment]
Ok. I'm so stupid... I had ditched the "--universal" for the build so, of course, the filename was different. I'll move to our scripts later. Only issue left is authorisation to DockerHub. That should be solvable... I'm confident I'll get this running tonight.

**Comment by menkej** [comment]
Finally. Sometimes its just the little things that suck... It finally works how I planned! :-) Please feel free to rip apart #223, guys!
